File: lms/lms/report/lms_collections/lms_collections.py
import frappe
from frappe.utils import flt


def execute(filters=None):
	filters = filters or {}
	columns = get_columns()
	data    = get_data(filters)
	summary = get_summary(data)
	return columns, data, None, None, summary


def get_columns():
	return [
		{"label": "Customer",              "fieldname": "customer",          "fieldtype": "Link",    "options": "Customer", "width": 210},
		{"label": "Contracts",             "fieldname": "contract_count",    "fieldtype": "Int",                            "width": 100},
		{"label": "Total Invoiced (TZS)",  "fieldname": "total_invoiced",    "fieldtype": "Float",                          "width": 190},
		{"label": "Total Paid (TZS)",      "fieldname": "total_paid",        "fieldtype": "Float",                          "width": 170},
		{"label": "Outstanding (TZS)",     "fieldname": "total_outstanding", "fieldtype": "Float",                          "width": 180},
		{"label": "Collection Rate %",     "fieldname": "collection_rate",   "fieldtype": "Percent",                        "width": 150},
	]


def get_data(filters):
	conditions = [
		"pc.docstatus = 1",
		"pc.contract_status IN ('Ongoing', 'Completed')",
	]
	if filters.get("customer"):
		conditions.append("pc.customer = %(customer)s")
	if filters.get("from_date"):
		conditions.append("pc.contract_date >= %(from_date)s")
	if filters.get("to_date"):
		conditions.append("pc.contract_date <= %(to_date)s")

	where = " AND ".join(conditions)

	rows = frappe.db.sql(f"""
		SELECT
			pc.customer,
			COUNT(pc.name)            AS contract_count,
			SUM(pc.selling_price)     AS total_invoiced,
			SUM(pc.total_paid)        AS total_paid,
			SUM(pc.total_outstanding) AS total_outstanding
		FROM `tabPlot Contract` pc
		WHERE {where}
		GROUP BY pc.customer
		ORDER BY total_paid DESC
	""", filters, as_dict=True)

	data = []
	for row in rows:
		invoiced = flt(row.total_invoiced)
		paid     = flt(row.total_paid)
		rate     = (paid / invoiced * 100) if invoiced else 0
		data.append({
			"customer":          row.customer,
			"contract_count":    row.contract_count,
			"total_invoiced":    invoiced,
			"total_paid":        paid,
			"total_outstanding": flt(row.total_outstanding),
			"collection_rate":   rate,
		})
	return data


def get_summary(data):
	if not data:
		return []
	total_invoiced    = sum(flt(r["total_invoiced"])    for r in data)
	total_paid        = sum(flt(r["total_paid"])        for r in data)
	total_outstanding = sum(flt(r["total_outstanding"]) for r in data)
	overall_rate      = (total_paid / total_invoiced * 100) if total_invoiced else 0
	return [
		{"label": "Total Invoiced",          "value": total_invoiced,            "datatype": "Float", "indicator": "Blue"},
		{"label": "Total Collected",         "value": total_paid,                "datatype": "Float", "indicator": "Green"},
		{"label": "Total Outstanding",       "value": total_outstanding,         "datatype": "Float", "indicator": "Red"},
		{"label": "Overall Collection Rate", "value": round(overall_rate, 1),    "datatype": "Float", "indicator": "Blue"},
	]

File: lms/lms/report/lms_collections/lms_collections.js
frappe.query_reports["LMS Collections"] = {
	filters: [
		{
			fieldname: "customer",
			label: "Customer",
			fieldtype: "Link",
			options: "Customer"
		},
		{
			fieldname: "from_date",
			label: "From Date",
			fieldtype: "Date"
		},
		{
			fieldname: "to_date",
			label: "To Date",
			fieldtype: "Date",
			default: frappe.datetime.get_today()
		}
	],
	formatter: function (value, row, column, data, default_formatter) {
		let formatted = default_formatter(value, row, column, data);

		if (!data) {
			return formatted;
		}

		if (column.fieldname === "customer") {
			return `<span style="font-weight:600;color:#2f4f4f;">${formatted}</span>`;
		}

		if (column.fieldname === "total_paid") {
			return `<span style="font-weight:700;color:#2f9e44;">${formatted}</span>`;
		}

		if (column.fieldname === "total_outstanding") {
			return `<span style="font-weight:700;color:#e03131;">${formatted}</span>`;
		}

		if (column.fieldname === "collection_rate") {
			const rate = Number(data.collection_rate || 0);
			const color = rate >= 70 ? "#2f9e44" : rate >= 40 ? "#f08c00" : "#e03131";
			return `<span style="font-weight:700;color:${color};">${formatted}</span>`;
		}

		return formatted;
	}
};
